cpp/tests/utilities/column_wrapper.cuh

```diff
@@ -0,0 +1,396 @@
+     * treated as equal to other nulls. Defaults to true.
+     *---------------------------------------------------------------------------**/
+    __host__ __device__ elements_equal(gdf_column lhs, gdf_column rhs,
+                                       bool nulls_are_equal = true)
+        : lhs_col{lhs}, rhs_col{rhs}, nulls_are_equivalent{nulls_are_equal} {}
+
+    __device__ bool operator()(gdf_index_type row) {
+      bool const lhs_is_valid{gdf_is_valid(lhs_col.valid, row)};
+      bool const rhs_is_valid{gdf_is_valid(rhs_col.valid, row)};
+
+      if (lhs_is_valid && rhs_is_valid) {
+        return static_cast<ColumnType const*>(lhs_col.data)[row] ==
+               static_cast<ColumnType const*>(rhs_col.data)[row];
+      }
+
+      // If one value is valid but the other is not
+      if (lhs_is_valid xor rhs_is_valid) {
+        return false;
+      }
+
+      return nulls_are_equivalent;
+    }
+  };
+
+  /**---------------------------------------------------------------------------*
+   * @brief Compares if another column_wrapper is equal to this wrapper.
+   *
+   * Treats NULL == NULL
+   *
+   * @param rhs  The other column_wrapper to check for equality
+   * @return true The two columns are equal
+   * @return false The two columns are not equal
+   *---------------------------------------------------------------------------**/
+  bool operator==(column_wrapper<ColumnType> const& rhs) const {
+    if (the_column.size != rhs.the_column.size) return false;
+    if (the_column.dtype != rhs.the_column.dtype) return false;
+    if (the_column.null_count != rhs.the_column.null_count) return false;
+    if (the_column.dtype_info.time_unit != rhs.the_column.dtype_info.time_unit)
+      return false;
+
+    if (!(the_column.data && rhs.the_column.data))
+      return false;  // if one is null but not both
+
+    if (not thrust::all_of(rmm::exec_policy()->on(0),
+                           thrust::make_counting_iterator(0),
+                           thrust::make_counting_iterator(the_column.size),
+                           elements_equal{the_column, rhs.the_column})) {
+      return false;
+    }
+
+    CUDA_RT_CALL(cudaPeekAtLastError());
+
+    return true;
+  }
+
+ private:
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates and initializes the underyling gdf_column with host data.
+   *
+   * Creates a gdf_column and copies data from the host for it's data and
+   * bitmask. Sets the corresponding dtype based on the column_wrapper's
+   * ColumnType.
+   *
+   * @param host_data The vector of host data to copy to device for the column's
+   * data
+   * @param host_bitmask Optional vector of host data for the column's bitmask
+   * to copy to device
+   *---------------------------------------------------------------------------**/
+  void initialize_with_host_data(
+      std::vector<ColumnType> const& host_data,
```

Comment:
Not a vector. Template this function and use iterators. If you really want to - have another variant which takes a `gsl::span` / `std::span`.

Reply:
It certainly _is_ a vector, or rather I'm using a vector (both `std` and `thrust`) as convenient data structures for holding the column's data and bitmask buffers. 

I think it's absolutely valid to request a constructor that uses iterators, but I'm going to say in the interest of time that this will be future work.